src/components/query-list/query-list.js

```diff
@@ -0,0 +1,110 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import classnames from 'classnames/bind';
+
+import styles from './query-list.css';
+import ScrollView from '../scroll-view';
+import Impagination from '../impagination';
+
+const cx = classnames.bind(styles);
+
+export default class QueryList extends Component {
+  static propTypes = {
+    type: PropTypes.string.isRequired,
+    page: PropTypes.number,
+    pageSize: PropTypes.number,
+    loadHorizon: PropTypes.number,
+    itemHeight: PropTypes.number,
+    notFoundMessage: PropTypes.string,
+    collection: PropTypes.object.isRequired,
+    fetch: PropTypes.func.isRequired,
+    renderItem: PropTypes.func.isRequired,
+    onPage: PropTypes.func
+  };
+
+  static defaultProps = {
+    page: 1,
+    pageSize: 25,
+    notFoundMessage: 'Not Found'
+  }
+
+  state = {
+    readOffset: (this.props.page - 1) * this.props.pageSize
+  };
+
+  handleUpdate = (readOffset) => {
+    let { pageSize, onPage } = this.props;
+    let page = Math.floor(readOffset / pageSize) + 1;
+
+    if (page !== this.props.page) {
+      onPage(page);
+    }
+
+    if (readOffset !== this.state.readOffset) {
+      this.setState({ readOffset });
+    }
+  }
+
+  render() {
+    let {
+      type,
+      page,
+      pageSize,
+      loadHorizon,
+      notFoundMessage,
+      collection,
+      fetch,
+      itemHeight,
+      renderItem
+    } = this.props;
+    let {
+      readOffset
+    } = this.state;
+
+    let listLength = collection.isLoaded ? collection.length : page * pageSize;
+    let totalPages = Math.ceil(listLength / pageSize);
```

Comment:
If possible, I'd like to push all of this down into the state that is managed by the `Impagination` component, because things like length and total pages really are properties of the dataset. And should be inferrable. If we can't push them down, then we need to ask ourselves how we can improve impagination so that we can.